### PR TITLE
SOL-150690: Add listen_address server config with loopback default fo…

### PR DESCRIPTION
**Pull request by solace-awaheed**
…r non-OAuth client auth modes

## Summary

Adds a listen_address config field so the dev auth modes (disabled/static) bind to 127.0.0.1 by default, and refuses a non-loopback bind under mode: disabled unless the operator explicitly sets allow_remote_unauthenticated: true. Makes the unauthenticated dev modes safe by construction rather than by warning.

Fixes # SOL-150690

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Dependency update

## Motivation and Context

The server always bound all interfaces (fmt.Sprintf(":%d", cfg.Port)) with no way to restrict it. Combined with mcp_client_auth.mode: disabled — a legal local-dev config with no client authentication — anyone on the network could reach the server, which then executes their requests against the broker using the admin credential it holds. A dev convenience became an unauthenticated, network-reachable admin path to the broker, mitigated only by a WARN banner and operator discipline.

## Changes Made

- New listen_address config field (top-level, alongside port:) controlling the bind host.
- Auth-mode-derived default: unset listen_address resolves to 127.0.0.1 for disabled/static, and to all-interfaces (:port) for oauth (preserving current production behaviour).
- New allow_remote_unauthenticated override: disabled mode refuses a non-loopback bind at startup unless this is explicitly set; the error explains the risk. static keeps the loopback default but may bind non-loopback without the override (it is token-gated).
- Startup banner logs the effective bind address alongside the auth mode, for every mode.
- BindAddress() uses net.JoinHostPort so IPv6 hosts (::1 → [::1]:9090) format correctly for net.Listen.
- Docs updated: docs/configuration.md, docs/authentication.md, broker-config.example.yaml.

## Testing

**Test Environment:**
- Go version: go1.26.1
- OS: 14.6.1 (darwin 23.6.0)
- Broker version (if applicable):

**Tests Performed:**
- [x] Unit tests added/updated
- [ ] Integration tests added/updated (if applicable)
- [ ] E2E tests added/updated (if applicable)
- [x] Tested locally with `go test -race ./...`
- [x] Tested with real Solace broker

**Test Coverage:**
- Default resolution matrix: disabled/static → 127.0.0.1; oauth → all-interfaces; explicit value preserved; mixed-case mode;
  whitespace trimmed.
- Validation matrix: disabled + non-loopback refused (asserts risk wording + field name); override allows; loopback/localhost/LOCALHOST/IPv6 ::1 allowed; IPv6 :: refused; static/oauth non-loopback allowed; malformed address rejected.
- BindAddress() formatting incl. IPv6 bracketing; banner logs bind address for all three modes.
- Manual end-to-end against a running server: lsof confirms 127.0.0.1:port (dev default) vs *:port (override); refusal path leaves 0 listeners; oauth unchanged.

## Breaking Changes

**Impact**: Operators currently running mode: disabled or mode: static with no listen_address and relying on network reachability will find the server bound to 127.0.0.1 only after upgrade. This is intentional — that configuration was the vulnerability.

**Migration Guide**: To restore network binding, set listen_address explicitly (e.g. 0.0.0.0). Under mode: disabled this additionally requires allow_remote_unauthenticated: true to acknowledge the unauthenticated exposure. The recommended path for network-facing deployments remains mode: oauth, whose default is unchanged.

### Code Quality
- [x] Code follows the [coding standards](../.github/CONTRIBUTING.md#coding-standards)
- [x] Self-review completed (checked for typos, logic errors, edge cases)
- [x] Code is well-commented (explains "why", not "what")
- [x] Complex logic has explanatory comments
- [x] No commented-out code or debug statements left in

### Security
- [x] **No credentials in code** (checked all files, commits, and logs)
- [x] Followed [secure logging rules](../docs/secure-logging-rules.md)
- [x] Credential-carrying types implement `slog.LogValuer`
- [ ] No security vulnerabilities introduced (checked with golangci-lint/gosec)

### Testing
- [x] All tests pass locally: `go test -race ./...`
- [x] No race conditions detected
- [x] Edge cases covered by tests
- [x] Error paths tested
- [x] Test names clearly describe what is being tested

### Documentation
- [ ] README.md updated (if user-facing changes)
- [x] docs/ updated (if architecture/design changes)
- [x] godoc comments added/updated for exported functions
- [ ] CHANGELOG.md updated (if applicable)
- [x] Examples updated (if applicable)

### Git & CI
- [ ] All commits are signed off (DCO: `git commit -s`)
- [ ] Commits have clear, descriptive messages
- [ ] Branch is up to date with main (rebased if needed)
- [ ] No merge conflicts
- [ ] CI checks passing (lint, build, test, E2E)

### Breaking Changes (if applicable)
- [x] Breaking changes documented in commit message
- [x] Breaking changes documented in PR description
- [x] Migration guide provided
- [ ] Deprecated functionality marked with comments and deprecation notices

## Related Issues/PRs

- Related to #
- Depends on #
- Blocks #

---

**For Reviewers:**

**Focus Areas**: 

**Questions**:

### CODE REVIEW

**Review by solace-aross — COMMENTED** [review]
Strong change. This closes a real unauthenticated-network-exposure hole and does it "safe by construction" rather than by warning banner. The `disabled`-mode refusal with an explicit `allow_remote_unauthenticated` acknowledgment is the right shape, the error names both the risk and the three escape hatches, and the default resolution correctly runs case-insensitively before mode normalization. Test coverage is above the bar (default matrix, validation, IPv6 bracketing, case-insensitivity, whitespace).

One design question worth a look before merge.

**`static` mode can put the shared dev token on the wire in cleartext**

The guard intentionally exempts `static`: a non-loopback bind is allowed with no override because "it still authenticates every request against the shared dev token." That holds only if the transport is encrypted. Server-side TLS is optional here (`tls_cert_file`/`tls_key_file`, both-or-neither, no enforcement), so `mode: static` + `listen_address: 0.0.0.0` + no cert sends a long-lived shared bearer token over plaintext HTTP on a routable interface. Sniff it once and you have the same broker-admin-backed access the `disabled` guard was built to prevent.

The asymmetry is what stands out: `disabled` forces an explicit `allow_remote_unauthenticated: true`, while `static` silently permits the network bind with no acknowledgment and no TLS check. Did you weigh requiring TLS (or at least a startup WARN) when `static` binds a non-loopback address without a cert? A WARN would keep the dev ergonomics while removing the "looks authenticated, actually cleartext" trap. Your call. I'd at least document the cleartext caveat next to the `static`/network line in `docs/authentication.md`.

Minor: `net.ParseIP` is evaluated up to three times across `isLoopbackHost` + the form check on the validation path. Startup-only, so cosmetic. Flagging only so it's a conscious skip.

File: internal/banner/banner.go [hunk 1]
```diff
@@ -78,6 +83,7 @@ func LogStartupAuthMode(mode, issuer string) {
 	case "oauth":
 		slog.Info("MCP client auth: OAuth/OIDC", slog.String("issuer", issuer))
 	}
+	slog.Info("MCP server bind address", slog.String("listen_address", bindAddr))
 }
```
[the remark this answers is not in the document]
Comment (solace-aross) [reply internal/banner/banner.go:87]: +1 on this. The value logged is the effective host:port but the config field `listen_address` is host-only, so `bind_address` reads truer for log search and alerting. One-word fix worth taking (update the banner tests/docs that assert the key).

[the remark this answers is not in the document]
Comment (solace-awaheed) [reply internal/banner/banner.go:87]: Updating, thanks!

**Review by solace-aross — APPROVED** [review]
Second pass, and it's ready. Approving.

The `static` cleartext concern from the first round is fully addressed, and addressed at the right layer: `StaticTokenExposedCleartext()` (static + non-loopback + no TLS) drives a startup WARN, the caveat is documented next to the `static`/network line in `docs/authentication.md`, and both the predicate and the banner have tests. Keeping it a WARN rather than a hard error is the right call — the network bind is a deliberate dev choice and TLS at a load balancer is a legitimate setup.

The `bind_address` log-key rename landed too, so the logged host:port no longer reads as the host-only `listen_address` field.

Nice work. The guardrail is safe-by-construction, the error names the risk and all three escape hatches, and coverage (default matrix, validation, IPv6 bracketing, case-insensitivity, cleartext predicate) is above the bar. CI is green across build, lint, and all three E2E suites.
